### ISSUE TITLE
Serialization hangs randomly

### ISSUE THREAD

**Issue by CaptainJiNX** [rkyv/rkyv#555]
When I run the code (with the provided json data) in:

 https://gist.github.com/CaptainJiNX/a9bc6ba3c07b44ccd75452af0e241638 
 
It sometimes completes the serialization, and sometimes it just hangs and gets stuck. Looking at the call stack when debugging and pausing the execution, it seems to be related to something in `ArchivedHashTable<T>::serialize_from_iter`

```
rkyv::simd::neon::Group::unpack (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/simd/neon.rs:68)
rkyv::simd::neon::Group::match_empty (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/simd/neon.rs:80)
rkyv::collections::swiss_table::table::ArchivedHashTable<T>::serialize_from_iter::{{closure}}::{{closure}} (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/table.rs:418)
rkyv::util::ser_vec::SerVec<T>::with_capacity (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/util/ser_vec.rs:56)
rkyv::collections::swiss_table::table::ArchivedHashTable<T>::serialize_from_iter::{{closure}} (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/table.rs:389)
rkyv::util::ser_vec::SerVec<T>::with_capacity (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/util/ser_vec.rs:56)
rkyv::collections::swiss_table::table::ArchivedHashTable<T>::serialize_from_iter (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/table.rs:379)
rkyv::collections::swiss_table::map::ArchivedHashMap<K,V,H>::serialize_from_iter (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/map.rs:220)
rkyv::impls::std::collections::hash_map::<impl rkyv::traits::Serialize<S> for std::collections::hash::map::HashMap<K,V,RandomState>>::serialize (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/impls/std/collections/hash_map.rs:40)
<rkyv_test::Item as rkyv::traits::Serialize<__S>>::serialize (/Users/roger/projects/tmp/rust/rkyv-test/src/main.rs:19)
<rkyv::collections::util::EntryAdapter<BK,BV,K,V> as rkyv::traits::Serialize<S>>::serialize (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/util.rs:70)
rkyv::collections::swiss_table::table::ArchivedHashTable<T>::serialize_from_iter::{{closure}}::{{closure}}::{{closure}} (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/table.rs:469)
rkyv::util::ser_vec::SerVec<T>::with_capacity (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/util/ser_vec.rs:56)
rkyv::collections::swiss_table::table::ArchivedHashTable<T>::serialize_from_iter::{{closure}}::{{closure}} (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/table.rs:460)
rkyv::util::ser_vec::SerVec<T>::with_capacity (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/util/ser_vec.rs:56)
rkyv::collections::swiss_table::table::ArchivedHashTable<T>::serialize_from_iter::{{closure}} (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/table.rs:389)
rkyv::util::ser_vec::SerVec<T>::with_capacity (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/util/ser_vec.rs:56)
rkyv::collections::swiss_table::table::ArchivedHashTable<T>::serialize_from_iter (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/table.rs:379)
rkyv::collections::swiss_table::map::ArchivedHashMap<K,V,H>::serialize_from_iter (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/collections/swiss_table/map.rs:220)
rkyv::impls::std::collections::hash_map::<impl rkyv::traits::Serialize<S> for std::collections::hash::map::HashMap<K,V,RandomState>>::serialize (/Users/roger/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rkyv-0.8.5/src/impls/std/collections/hash_map.rs:40)
```

Not sure what's going on there...

**Comment by djkoloski** [comment]
I made the repro 100% consistent by explicitly setting the hasher:

```rust
#[derive(rkyv::Archive, rkyv::Serialize, serde::Deserialize)]
struct Index {
    by_id: HashMap<String, Item, BuildHasherDefault<FxHasher64>>,
}

#[derive(rkyv::Archive, rkyv::Serialize, serde::Deserialize)]
struct Item {
    by_key: HashMap<String, String, BuildHasherDefault<FxHasher64>>,
}
```

Now I can dig in a little further.

**Comment by djkoloski** [comment]
Fixed in 654c3eb65995515fb9a0c9057c7645103bf3341b and released in 0.8.6. This was a bit gnarly.

rkyv contains a portable implementation of swiss tables, but makes a few changes for conserving space on the wire. One of those changes is that the number of control bytes is reduced to a multiple of 16 rather than a power of two. This is because the size of the hash map won't increase later, so this can save a significant amount of space for some hash maps.

As a result, rkyv has a slightly different probing algorithm than swiss tables uses. This issue was caused by an incorrect assumption that we can skip any probes that aren't within the _capacity_ of the hash table. This didn't turn out to be the case for very technical reasons and now hash maps need more control bytes. Unfortunately this is a format breaking change, but this issue can't be fixed any other way without severely hampering performance.

Thanks for providing a complete repro, I don't think I would have been able to track this down very quickly or easily otherwise.
